devito/finite_differences/derivative.py

```diff
@@ -0,0 +1,192 @@
+        elif len(dims) == 1:
+            if isinstance(dims[0], Iterable):
+                # Iterable of Dimensions
+                if len(dims[0]) != 2:
+                    raise ValueError("Expected `(dim, deriv_order)`, got %s" % dims[0])
+                orders = kwargs.get('deriv_order', dims[0][1])
+                if dims[0][1] != orders:
+                    raise ValueError("Two different values of `deriv_order`")
+                new_dims = tuple([dims[0][0]]*dims[0][1])
+            else:
+                # Single Dimension
+                orders = kwargs.get('deriv_order', 1)
+                if isinstance(orders, Iterable):
+                    orders = orders[0]
+                new_dims = tuple([dims[0]]*orders)
+        else:
+            # Iterable of 2-tuple, e.g. ((x, 2), (y, 3))
+            new_dims = []
+            orders = []
+            d_ord = kwargs.get('deriv_order', tuple([1]*len(dims)))
+            for d, o in zip(dims, d_ord):
+                if isinstance(d, Iterable):
+                    new_dims.extend([d[0] for _ in range(d[1])])
+                    orders.append(d[1])
+                else:
+                    new_dims.extend([d for _ in range(o)])
+                    orders.append(o)
+            new_dims = as_tuple(new_dims)
+            orders = as_tuple(orders)
+
+        # Finite difference orders
+        fd_orders = kwargs.get('fd_order', tuple([expr.space_order]*len(dims)))
+        if len(dims) == 1 and isinstance(fd_orders, Iterable):
+            fd_orders = fd_orders[0]
+        # SymPy expects the list of variable w.r.t. which we differentiate to be a list
+        # of 2-tuple `(s, count)` where s is the entity to diff wrt and count is the order
+        # of the derivative
+        variable_count = [sympy.Tuple(s, new_dims.count(s))
+                          for s in filter_ordered(new_dims)]
+
+        # Construct the actual Derivative object
+        obj = Differentiable.__new__(cls, expr, *variable_count)
+        obj._dims = tuple(OrderedDict.fromkeys(new_dims))
+        obj._fd_order = fd_orders
+        obj._deriv_order = orders
+        obj._side = kwargs.get("side", centered)
+        obj._stagger = kwargs.get("stagger", tuple([left]*len(obj._dims)))
+        obj._transpose = kwargs.get("transpose", direct)
+
+        return obj
+
+    @cached_property
+    def _args_diff(self):
+        return tuple(i for i in self.args if isinstance(i, Differentiable))
+
+    @property
+    def dims(self):
+        return self._dims
+
+    @property
+    def fd_order(self):
+        return self._fd_order
+
+    @property
+    def deriv_order(self):
+        return self._deriv_order
+
+    @property
+    def stagger(self):
+        return self._stagger
+
+    @property
+    def side(self):
+        return self._side
+
+    @property
+    def transpose(self):
+        return self._transpose
+
+    @property
+    def T(self):
+        """Transpose of the Derivative."""
```

Comment:
this is clear to me, but I'm not sure the average user will understand what we exactly mean by "transpose of the derivative" ?

Reply:
Added more explanation